src/cpp/rtps/builtin/data/ParticipantProxyData.cpp

```diff
@@ -95,339 +95,338 @@ ParticipantProxyData::~ParticipantProxyData()
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        ParameterString_t p(PID_ENTITY_NAME, 0, m_participantName);
+        if (!p.addToCDRMessage(msg)) return false;
     }
 
     if(this->m_userData.size()>0)
     {
-        UserDataQosPolicy* p = new UserDataQosPolicy();
-        p->setDataVec(m_userData);
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        UserDataQosPolicy p;
+        p.setDataVec(m_userData);
+        if (!p.addToCDRMessage(msg)) return false;
     }
 
     if(this->m_properties.properties.size()>0)
     {
-        ParameterPropertyList_t* p = new ParameterPropertyList_t(m_properties);
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        ParameterPropertyList_t p(m_properties);
+        if (!p.addToCDRMessage(msg)) return false;
     }
 
 #if HAVE_SECURITY
     if(!this->identity_token_.class_id().empty())
     {
-        ParameterToken_t* p = new ParameterToken_t(PID_IDENTITY_TOKEN, 0);
-        p->token = identity_token_;
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        ParameterToken_t p(PID_IDENTITY_TOKEN, 0);
+        p.token = identity_token_;
+        if (!p.addToCDRMessage(msg)) return false;
     }
 
     if(!this->permissions_token_.class_id().empty())
     {
-        ParameterToken_t* p = new ParameterToken_t(PID_PERMISSIONS_TOKEN, 0);
-        p->token = permissions_token_;
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        ParameterToken_t p(PID_PERMISSIONS_TOKEN, 0);
+        p.token = permissions_token_;
+        if (!p.addToCDRMessage(msg)) return false;
     }
 
     if ((this->security_attributes_ != 0UL) || (this->plugin_security_attributes_ != 0UL))
     {
-        ParameterParticipantSecurityInfo_t* p = new ParameterParticipantSecurityInfo_t();
-        p->security_attributes = this->security_attributes_;
-        p->plugin_security_attributes = this->plugin_security_attributes_;
-        parameter_list.m_parameters.push_back((Parameter_t*)p);
+        ParameterParticipantSecurityInfo_t p;
+        p.security_attributes = this->security_attributes_;
+        p.plugin_security_attributes = this->plugin_security_attributes_;
+        if (!p.addToCDRMessage(msg)) return false;
     }
 #endif
 
-    return parameter_list;
+    return CDRMessage::addParameterSentinel(msg);
 }
 
 bool ParticipantProxyData::readFromCDRMessage(CDRMessage_t* msg, bool use_encapsulation)
 {
-    ParameterList_t parameter_list;
-
-    if(ParameterList::readParameterListfromCDRMsg(msg, &parameter_list, NULL, use_encapsulation) > 0)
+    auto param_process = [this](const Parameter_t* param)
     {
-        for(std::vector<Parameter_t*>::iterator it = parameter_list.m_parameters.begin();
-                it!=parameter_list.m_parameters.end();++it)
+        switch (param->Pid)
         {
-            switch((*it)->Pid)
+            case PID_KEY_HASH:
+            {
+                ParameterKey_t*p = (ParameterKey_t*)(param);
```

Comment:
Better use dynamic_cast instead of c-style casting.

Reply:
done